### PR TITLE
Frequency filtering

### PR DESCRIPTION
**Pull request by djesic**

### CODE REVIEW

File: src/RegisterMapping.h [hunk 1]
```diff
@@ -85,9 +85,11 @@ class RegisterMapping
      * @param readRestricted indicates whether or not the mapping can be read
      * @param slaveAddress of the devices being accessed, leave as default on -1
      * @param deadbandValue indicates a change in value of the register that is insignificant data
+     * @param frequencyFilterValue changes that occur within the given time (in miliseconds) that will be ignored
      */
     RegisterMapping(const std::string& reference, RegisterType registerType, int32_t address,
-                    bool readRestricted = false, int16_t slaveAddress = -1, double deadbandValue = 0.0);
+                    bool readRestricted = false, int16_t slaveAddress = -1, double deadbandValue = 0.0,
+                    unsigned long long frequencyFilterValue = 0);
```
Comment (milanzlicic) [inline src/RegisterMapping.h]: this should be of type std::chrono::milliseconds

File: src/RegisterMapping.h [hunk 2]
```diff
@@ -237,6 +246,8 @@ class RegisterMapping
     bool m_isInitialized = false;
     bool m_isValid = false;
     double m_deadbandValue = 0.0;
+    unsigned long long m_lastUpdateTime = 0;
+    unsigned long long m_frequencyFilterValue = 0;
```
Comment (milanzlicic) [inline src/RegisterMapping.h]: std::chrono::milliseconds

File: src/RegisterMapping.h [hunk 3]
```diff
@@ -237,6 +246,8 @@ class RegisterMapping
     bool m_isInitialized = false;
     bool m_isValid = false;
     double m_deadbandValue = 0.0;
+    unsigned long long m_lastUpdateTime = 0;
```
Comment (milanzlicic) [inline src/RegisterMapping.h]: std::chrono::time_point\<std::chrono::high_resolution_clock\>

File: src/RegisterMapping.h [hunk 4]
```diff
@@ -132,13 +138,16 @@ class RegisterMapping
...
+     * @param frequencyFilterValue changes that occur within the given time (in miliseconds) that will be ignored
      */
     RegisterMapping(const std::string& reference, RegisterType registerType, const std::vector<int32_t>& addresses,
                     OutputType type, OperationType operation, bool readRestricted = false, int16_t slaveAddress = -1,
-                    double deadbandValue = 0.0);
+                    double deadbandValue = 0.0, unsigned long long frequencyFilterValue = 0);
 
     virtual ~RegisterMapping() = default;
 
+    static unsigned long long int currentRtc();
```
Comment (milanzlicic) [inline src/RegisterMapping.h]: Once you replace types of filter and lastUpdateTime there is no need for this method
Just call std::chrono::time_point\<std::chrono::high_resolution_clock\>::now() where needed

**Review by milanzlicic — CHANGES_REQUESTED** [review]

File: src/RegisterMapping.cpp [hunk 5]
```diff
@@ -260,104 +275,51 @@ bool RegisterMapping::doesUpdate(const std::vector<uint16_t>& newValues) const
...
-    {
-        return !m_isInitialized || !m_isValid;
-    }
-
-    bool significantChange = false;
-
-    switch (getOutputType())
-    {
-    case OutputType::UINT16:
+    if (!m_isInitialized)
```
Comment (milanzlicic) [inline src/RegisterMapping.cpp]: Do not ignore m_isValid. If it is false you should also return true

File: src/RegisterMapping.cpp [hunk 6]
```diff
@@ -385,11 +347,25 @@ bool RegisterMapping::update(const std::vector<uint16_t>& newValues)
...
     m_isValid = true;
 
+    m_lastUpdateTime = std::chrono::high_resolution_clock::now();
+
     return !isValueInitialized || different || !isValid;
 }
 
 bool RegisterMapping::doesUpdate(bool newRegisterValue) const
 {
+    if (!m_isInitialized)
```
Comment (milanzlicic) [inline src/RegisterMapping.cpp]: As above, do not ignore m_isValid. It is used to report value when the previous read has failed

File: src/RegisterMapping.cpp [hunk 7]
```diff
@@ -260,104 +275,51 @@ bool RegisterMapping::doesUpdate(const std::vector<uint16_t>& newValues) const
...
-    case OutputType::UINT32:
-    {
-        uint32_t currentValueUint32;
-        uint32_t newValueUint32;
-        if (getOperationType() == OperationType::MERGE_BIG_ENDIAN)
+        if (m_deadbandValue == 0.0)
         {
-            currentValueUint32 = DataParsers::registersToUint32(m_byteValues, DataParsers::Endian::BIG);
-            newValueUint32 = DataParsers::registersToUint32(newValues, DataParsers::Endian::BIG);
+            return !frequentUpdate;    // freq filter
```
Comment (milanzlicic) [inline src/RegisterMapping.cpp]: Will this not return true even if no change in register value?
Mayle you should move the "if (!different)" check from below to above checking freq so you only check if there is actually diff in values

File: tests/ComplexMappingsTests.cpp [hunk 8]
```diff
@@ -280,6 +282,37 @@ TEST_F(ComplexMappingsTests, UInt32MappingsDeadband)
...
+        EXPECT_FALSE(mapping->isValid());
+
+        EXPECT_NO_THROW(mapping->update(bytes));
+        EXPECT_EQ(value, mapping->getUint32Value());
+
+        EXPECT_TRUE(mapping->isInitialized());
+        EXPECT_TRUE(mapping->isValid());
+
+        EXPECT_FALSE(mapping->doesUpdate(wolkabout::DataParsers::uint32ToRegisters(value + 1, endian)));
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Comment (milanzlicic) [inline tests/ComplexMappingsTests.cpp:311]: Instead of sleeping, try setting the saved value to -100 to avoid delay

**Review by milanzlicic — CHANGES_REQUESTED** [review]

File: tests/ComplexMappingsTests.cpp (hunk 8 above)
Comment (djesic) [reply tests/ComplexMappingsTests.cpp:311]: I'd need to write a setter that will only be used in tests.
Currently, this test increases the total test duration from 8.8s to 9.6s
Do I still write it?

Comment (milanzlicic) [reply tests/ComplexMappingsTests.cpp:311]: You'd could actually override the access modifiers of class by writting "#define private public" before including Register mapping header, and thus being able to modify private members of class without the need for setters. We do it on other projects as well. But you may skip this for now as you only add delay on these two test cases, but it should be done if more delays are to be added

File: src/RegisterMapping.cpp (hunk 6 above)
Comment (milanzlicic) [reply src/RegisterMapping.cpp]: Still missing the m_isValid check as it will not reach the last return statement if rejected by freq filter

**Review by milanzlicic — CHANGES_REQUESTED** [review]

File: tests/ComplexMappingsTests.cpp (hunk 8 above)
Comment (djesic) [reply tests/ComplexMappingsTests.cpp:311]: Noted.
I'll leave it as it is right now, but will update if we add more tests.

**Review by milanzlicic — APPROVED** [review]

**Pull request merged**